### ISSUE TITLE
Issues when switching branches

### ISSUE THREAD

**Issue by akinsho** [lewis6991/gitsigns.nvim#44]
When switching git branches using `vim-fugitive` or `lazygit.nvim` I get a series of errors. When *just* switching branch I get the aforementioned `/lua/gitsigns.lua:663: attempt to index local 'bcache' (a nil value)`. However when opening the `lazygit` terminal floating window I get

![image](https://user-images.githubusercontent.com/22454918/103663184-b9f58300-4f68-11eb-8fe3-40cec2bf2d3e.png)
I can't tell if it's trying to perform git operations on this invalid buffer somehow, it just generally really seems to react very badly to opening that plugin window, but ignores others like FZF so could be `git` related or could be some buffer attribute of lazygit isn't set up properly

**Comment by cloggier** [comment]
I am experiencing the same. Thing it might have something to do with file(s) not existing on disk after branch change.

**Comment by lewis6991** [comment]
Could you guys try out #46. Hopefully it fixes this.

**Comment by akinsho** [comment]
Will give it a try and report back 👍🏾

**Comment by akinsho** [comment]
Been using this this morning and no errors so far switching git branches or using `lazygit` 👍🏿

**Comment by akinsho** [comment]
Hmm actually I just hit part of this again this morning, thankfully I'm not seeing the large job spawn error anymore, but I am still hitting the `bcache` issue. It doesn't seem to happen on a regular branch switch, I think in the case I noticed it I had some conflicts on rebasing a branch, not sure if that helps or not. I think 46 is definitely an improvement though

**Comment by lewis6991** [comment]
Ok thanks for the report. Shall we wait a few more days until you understand what scenario the issue happens under? I'm sure the fix for it will be trivial.

**Comment by akinsho** [comment]
Hmm after a few more days of use one thing that seems to be happening is that git signs is trying to track buffers that it shouldn't apply to and the `bcache` nil error is happening. Not sure of any of the underlying logic but I think if a buffer has a non empty `buftype` gitsigns should just not be active in that buffer.  

I'm getting these errors when I use fugitive buffers, and initially in the lazygit buffer but that maybe has stopped not 100% sure.

From this list of buftypes from the docs seems that none of them are good candidates to have git signs enabled in
```
	  acwrite	buffer will always be written with |BufWriteCmd|s
	  help		help buffer (do not set this manually)
	  nofile	buffer is not related to a file, will not be written
	  nowrite	buffer will not be written
	  quickfix	list of errors |:cwindow| or locations |:lwindow|
	  terminal	|terminal-emulator| buffer
	  prompt	buffer where only the last line can be edited, meant

```

**Comment by akinsho** [comment]
![gitsigns-stage-error](https://user-images.githubusercontent.com/22454918/104451687-ca75b100-5599-11eb-8e43-6a357e95d6b2.gif)

**Comment by lewis6991** [comment]
Can you provide a minimal example with precise instructions? I can't trigger any error.

**Comment by akinsho** [comment]
🤔 I will try to, it's a little intermittent and I can tell if it's the project as in a specific repository and something about it causing the errors, or the system since I use a mac for work and linux for my personal machine and the errors seem to happen almost, if not exclusively, on my mac but I'm not sure if that's just because that's when I work on the repo in question

**Comment by akinsho** [comment]
@lewis6991 sorry I've been quite busy the last few days and so haven't had a chance to sit and write up a minimal test case. I have been using your branch for a while now though and am pretty confident that the specific issue still relates to when moving across branches and files are added or removed when switching.

I believe you will be able to reproduce it in a simple repo by adding creating a new branch and adding a file to it, then moving between that and the main branch. At least this is when I encounter the error, it doesn't happen just if files change it seems specifically to relate to missing files

**Comment by lewis6991** [comment]
Does it matter how the branch is switched? Can this be done in another terminal whilst nvim is running?

**Comment by akinsho** [comment]
@lewis6991 I'm currently using `vim-fugitive` and `lazygit.nvim` to switch branches. Tbh I very rarely switch branches in the terminal proper I almost always use either of those tools. I just tried both and it reproduced using `fugitive` but not in the terminal (nvim's terminal not one external to nvim)

**Comment by cloggier** [comment]
I always switch branches in separate terminal and am getting that error, but didn't have time to check out that branch unfortunately.

**Comment by lewis6991** [comment]
I've merged the PR as it seems to improve things but will keep this issue open until we resolve all the known issues.

**Comment by akinsho** [comment]
@lewis6991 this might go without saying but I tried logging inside of `gitsigns.lua` for where I'm seeing the null error and `cache` is empty at this point. In this particular case I had just checked out a local version of a branch with various files open and had just merged in some upstream changes that tweaked quite a few files, I then fetched it and rebased my current branch onto the upstream branch which is when I hit this. I'm wondering if the fact that I have the following in my `gitconfig` is a factor.
```
[pull]
  rebase = true

```

**Comment by akinsho** [comment]
I'm still experiencing this, sorry it's been quiet in terms of reporting things back. I wonder if there is a debug mode I could run gitsigns in that I could then use to post a log for when next this happens. It typically happens in my work repo and much more rarely in my dotfiles repo but I'm usually to preoccupied at the time to stop and debug, but maybe if could check a log at some later point it might indicate something.

**Comment by lewis6991** [comment]
> EDIT: alternatively are there some steps I could use to investigate I'm not really sure what to look for as to why it might be happening e.g. logging at any specific points etc in my version of the plugin. From what I see and which is visible in the original screen shot. Plenary is trying to start a job in a folder that doesn't exist between switching branches. It seems not to be using my git repository as the root to use for it's jobs.

I have something like this:

```lua
vim.api.nvim_set_keymap('n', 'm', ':lua require"gitsigns".dump_cache()<cr>', {silent=true})
vim.api.nvim_set_keymap('n', 'M', ':lua require"gitsigns".debug_messages()<cr>', {silent=true})
require('gitsigns').setup{
  debug_mode = true
}
```

`debug_messages` just prints debug messages to `:messages`. Most of the messages are just when we run git commands or when we attach/detach from a buffer. You should be able to add your own using the `dprint` function. The cache is also useful to look at because it tells you everything that gitsigns has determined about the buffer. I bet the cwd for one of the buffers is getting messed up somehow.

It might be worth studying the `attach` function, it should be straight forward enough. We run a git command in the directory of the buffers filepath to figure out the git toplevel and store that in the cache to use as a cwd for future jobs. Maybe the issue lies around there?

**Comment by lewis6991** [comment]
@akinsho give #82 a go.

**Comment by akinsho** [comment]
No signs of the bug so far, but given the sporadic nature not sure if that means much. Will do some more testing/try triggering it by creating some fake nested dirs on different branches etc.

**Comment by lewis6991** [comment]
Well I know it fixes at least error I was able to reproduce (steps in PR).

**Comment by akinsho** [comment]
So far so good @lewis6991 not seen the error since running #44. I was thinking that for sanity sake maybe leaving this open till the end of the week and if there are no more appearances of the bug then maybe closing it out, can revisit if it returns.

Whilst I think the specific issue about switching branches might be resolved 🤞🏿 but there is another issue which is the `bcache <bufnum> is nil`
```
...e/pack/packer/start/gitsigns.nvim/lua/gitsigns/async.lua:53: ...im/site/pack/packer/start/gitsigns.nvim/lua/gitsigns.lua:137: Cache for buffer 9 was nil
```

dump cache shows after the error
```
  [9] = {
    abbrev_head = "main",
    file = "/Users/akinso/.dotfiles/.config/nvim/lua/as/plugins/gitsigns.lua",
    gitdir = "/Users/akinso/.dotfiles/.git",
    has_conflicts = false,
    hunks = {},
    index_watcher = <userdata 4>,
    mode_bits = "100644",
    object_name = "ba57c662aaeb35e8aecff567266852389b98e32e",
    pending_signs = {},
    relpath = ".config/nvim/lua/as/plugins/gitsigns.lua",
    staged = "/tmp/lua_7lEZRL",
    staged_diffs = {},
    toplevel = "/Users/akinso/.dotfiles"
  },
```

dump messages shows
```
watch_index(9): Watching index
dprint(nil): Running: git --no-pager show :0:.config/nvim/lua/as/plugins/gitsigns.lua
dprint(nil): Running: git --no-pager diff --color=never --diff-algorithm=histogram --patch-with-raw --unified=0 /tmp/lua_7lEZRL /tmp/lua_7lEZRL_buf
update(9): updates: 13, jobs: 72
watcher_cb(11): Index update
dprint(nil): Running: git rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
watcher_cb(2): Index update
dprint(nil): Running: git rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
watcher_cb(8): Index update
dprint(nil): Running: git rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
watcher_cb(4): Index update
dprint(nil): Running: git rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
watcher_cb(10): Index update
dprint(nil): Running: git rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
dprint(nil): Running: git --no-pager ls-files --stage --others --exclude-standard /Users/akinso/.dotfiles/.config/nvim/lua/as/plugins/init.lua
dprint(nil): Running: git --no-pager ls-files --stage --others --exclude-standard /Users/akinso/.dotfiles/.config/nvim/lua/as/highlights.lua
dprint(nil): Running: git --no-pager ls-files --stage --others --exclude-standard /Users/akinso/.dotfiles/.config/nvim/lua/as/plugins/telescope.lua
dprint(nil): Running: git --no-pager ls-files --stage --others --exclude-standard /Users/akinso/.dotfiles/.config/nvim/lua/as/statusline/utils.lua
dprint(nil): Running: git --no-pager ls-files --stage --others --exclude-standard /Users/akinso/.dotfiles/.config/nvim/syntax/NeogitStatus.vim
watcher_cb(11): File not changed
watcher_cb(2): File not changed
watcher_cb(4): File not changed
watcher_cb(8): File not changed
watcher_cb(10): File not changed
```

I think this is worth raising and resolving separately. I lumped it in here so it's been a little murky that they are separate probably. This issue seems to happen most frequently when I change a buffer in place using git, e.g. I run `git pull` which is configured via my git config to rebase the changes so the current buffer contents change and I think that blats the cache somehow despite the buffer being the same.

**Comment by lewis6991** [comment]
Thanks for the report.

The debug messages are pretty confusing and don't seem to make any sense. Apparently there is no cache for buffer 9 when in fact there is. The log also doesn't have any attach/detach messages. If the buffer cache gets cleared, that can only be caused by a detach event. Maybe the log messages were earlier ones and you didn't scroll to the later ones, or reverse and the detach messages were higher up?

I have hit a similar issue today however. If I change a file which is open in nvim via an external git command, gitsigns is able to handle and updates the buffer accordingly, however if I undo in the buffer, then that causes the `bcache nil` error. I inspected a bit further and the undo action was causing a detach event. I'm not sure if this is intended behaviour and *may* be a bug in `nvim_buf_attach()`. I'm inclined to believe it is an nvim bug since an attach event doesn't occur after the detach, which it should since the buffer is still loaded. Either way, this is a particularly tricky issue, but I have at least figured out some steps to reliable reproduce.

I reckon my above issue is closely related to yours since a `bcache nil` error can only occur from a detach event as already mentioned.

**Comment by akinsho** [comment]
Thanks for the explanation 👍🏿 glad you found a path to reproduce it. Yeah sorry the error happened at an inopportune time so I had to come back to it a few mins later and could easily have been the wrong bits that I grabbed, I'd done some other logging with a previous similar error. Regarding the undo, there were some PRs in nvim to fix treesitter related bugs that seemed to necessitate reseting something about the state of the buffer on things like undo or reading the buffer after coming back since both of those points were where treesitter errors were cropping up.

That could be related, or fresh red herring 😅
